torch/_higher_order_ops/auto_functionalize.py

```diff
@@ -120,33 +124,205 @@ def can_auto_functionalize(op: OperatorBase) -> bool:
+
+            output_kwargs[f"_{arg_name}_length"] = len(arg)
+            for i, elem in enumerate(arg):
+                serialize_single_view(
+                    f"_{arg_name}_{i}", elem, arg_to_base_index[arg_name][i]
+                )
+
+        elif isinstance(arg_type, (torch.TensorType, torch.OptionalType)):
+            serialize_single_view(
+                f"_{arg_name}",
+                input_kwargs[arg_name],
+                arg_to_base_index.get(arg_name, None),
+            )
+        else:
+            raise RuntimeError(f"Unsupported type {arg_type}")
+
+
+# read the serialized size(), stride(), storage_offset() and return them in a dict that maps each args to its ViewInfo.
+def deserialize_views_meta(arg_names, arg_types, input_kwargs, all_bases, pop_args):
+    def get_arg(name):
+        if pop_args:
+            return input_kwargs.pop(name)
+
+        return input_kwargs[name]
+
+    def deserialize_single_view(prefix):
+        base_index = get_arg(f"{prefix}_base_index")
+        if base_index is None:
+            return None
+        else:
+            size = get_arg(f"{prefix}_size")
+            stride = get_arg(f"{prefix}_stride")
+            storage_offset = get_arg(f"{prefix}_storage_offset")
+            return ViewInfo(
+                base_index, all_bases[base_index], size, stride, storage_offset
+            )
+
+    args_view_info: Dict[str, Any] = {}
+    for arg_name, arg_type in zip(arg_names, arg_types):
+        if isinstance(arg_type, torch.ListType):
+            length = get_arg(f"_{arg_name}_length")
+            if length is None:
+                # The whole list is None.
+                args_view_info[arg_name] = None
+            else:
+                args_view_info[arg_name] = [
+                    deserialize_single_view(f"_{arg_name}_{i}") for i in range(length)
+                ]
+
+        elif isinstance(arg_type, (torch.TensorType, torch.OptionalType)):
+            args_view_info[arg_name] = deserialize_single_view(f"_{arg_name}")
+        else:
+            raise RuntimeError(f"Unsupported type {arg_type}")
+    return args_view_info
+
+
 @auto_functionalized.py_impl(DispatchKey.CompositeExplicitAutograd)
 def auto_functionalized_dense(
     _mutable_op: OpOverload,
     _only_clone_these_tensors: Optional[Tuple[str, ...]] = None,
     **kwargs: Any,
 ) -> Tuple[Any, Tuple[Tensor, ...]]:
+    all_bases: List[Tensor] = kwargs.pop("_all_bases", [])
+    mutable_args_names, mutable_args_types = get_mutable_args(_mutable_op)
+    args_view_info = deserialize_views_meta(
+        mutable_args_names, mutable_args_types, kwargs, all_bases, pop_args=True
+    )
+
+    def regenerate_view(ViewInfo):
+        if (
+            ViewInfo.base.size() == ViewInfo.size
+            and ViewInfo.base.stride() == ViewInfo.stride
+            and ViewInfo.base.storage_offset() == ViewInfo.storage_offset
+        ):
+            return ViewInfo.base
```

Comment:
I'm worried the comparision might create additional dynamic shape guards

Reply:
mhmm you mean if ViewInfo.base.size(), ViewInfo.base.stride(), ViewInfo.base.storage_offset() were symbols and not constants.

Reply:
Yes

Reply:
removed with the new changes